src/ray/common/client_connection.cc

```diff
@@ -14,21 +14,50 @@
 
 #include "client_connection.h"
 
-#include <stdio.h>
-
 #include <boost/asio/buffer.hpp>
 #include <boost/asio/generic/stream_protocol.hpp>
 #include <boost/asio/placeholders.hpp>
 #include <boost/asio/read.hpp>
 #include <boost/asio/write.hpp>
 #include <boost/bind.hpp>
+#include <chrono>
 #include <sstream>
+#include <thread>
 
 #include "ray/common/ray_config.h"
 #include "ray/util/util.h"
 
 namespace ray {
 
+Status ConnectSocketRetry(local_stream_socket &socket, const std::string &endpoint,
+                          int num_retries, int64_t timeout_in_ms) {
+  RAY_CHECK(num_retries != 0);
+  // Pick the default values if the user did not specify.
+  if (num_retries < 0) {
+    num_retries = RayConfig::instance().raylet_client_num_connect_attempts();
+  }
+  if (timeout_in_ms < 0) {
+    timeout_in_ms = RayConfig::instance().raylet_client_connect_timeout_milliseconds();
+  }
+  boost::system::error_code ec;
+  for (int num_attempts = 0; num_attempts < num_retries; ++num_attempts) {
+    // The latest boost::asio always returns void for connect(). Do not
+    // treat its return value as error code anymore.
+    socket.connect(ParseUrlEndpoint(endpoint), ec);
+    if (!ec) {
+      break;
+    }
+    if (num_attempts > 0) {
+      RAY_LOG(ERROR) << "Retrying to connect to socket for endpoint " << endpoint
+                     << " (num_attempts = " << num_attempts
+                     << ", num_retries = " << num_retries << ")";
+    }
+    // Sleep for timeout milliseconds.
+    std::this_thread::sleep_for(std::chrono::milliseconds(timeout_in_ms));
+  }
+  return boost_to_ray_status(ec);
+}
+
 std::shared_ptr<ServerConnection> ServerConnection::Create(local_stream_socket &&socket) {
```

Comment:
Do you know if this function can be removed without much negative impact? It doesn't really seem to provide a useful abstraction.

Reply:
Do you mean "ConnectSocketRetry"? I find it ("ConnectSocketRetry") will be used later in 3 more places (it is originally used in Plasma Store). This is why I try to keep this function.

Reply:
Sorry no, I meant `ServerConnection::Create`. I realize you haven't modified it, I just mentioned it since it seems like it might be worth the cleanup after the surrounding refactor.

Reply:
Oh, this magic `ClientConnection::Create` is critical to maintain the lifetime of this object (it is extremely confusing, but I remember it is from boost official examples). One example is here:

https://github.com/ray-project/ray/blob/fcdf410ae1bb5071e7d92174eace7d79be2d4ef9/src/ray/raylet/raylet.cc#L129-L148

In this example, it looks like `new_connection` will go out of scope, and we are leaking memory. But actually `ClientConnection::Create` calls `client_handler`, and `client_handler` will call `ClientConnection::ProcessMessages()`, then `ClientConnection::ProcessMessages()` will general a `shared_ptr` of itself via `shared_ClientConnection_from_this()` in a boot::asio callback, so that its lifetime will be only controlled by `boost::asio`. This is quite tricky, but there is hardly other ways to achieve it.

Reply:
I agree with you that forcing `ServerConnection` to be a `shared_ptr` could be over decorating, but I don't have a strong mind to remove it.

Reply:
Sorry, I was just referring to `ServerConnection::Create`, not `ClientConnection::Create`. Does removing `ServerConnection::Create` have any impact on users of `ClientConnection::Create`? I was imagining it would only imply a trivial change to callers callers to use `std::make_shared`, but I'm not sure.

Reply:
(In any case I'm imagining this is a < 5-minute change, so just disregard my comment if it's more complicated.)

Reply:
I can do it in a later PR.